hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/ReceiptCache.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.state;
+
+import com.hedera.hapi.node.base.AccountID;
+import com.hedera.hapi.node.base.TransactionID;
+import com.hedera.hapi.node.transaction.TransactionReceipt;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.Set;
+
+/**
+ * A time-limited cache of transaction receipts.
+ *
+ * <p>Each {@link com.hedera.hapi.node.base.Transaction} has a unique {@link TransactionID}. Each {@link TransactionID}
+ * is valid for only a certain period of time. If submitted before the {@link TransactionID#transactionValidStart()}
+ * then the transaction is invalid. If submitted after the network configured expiration time from that start time,
+ * then the transaction is invalid. Between those two times, the network needs to deduplicate transactions.
+ *
+ * <p>It may be that the same transaction is submitted by the user multiple times to the same node. The node uses this
+ * cache as a mechanism to detect those duplicate transactions and reject the duplicates.
+ *
+ * <p>It may be that transactions with the same {@link TransactionID} may be submitted by the user to multiple nodes.
+ * If this happens, we need to only process one of those transactions, but charge the node + network fees to the user
+ * for each transaction they submitted.
+ *
+ * <p>It may be that a dishonest node receives a transaction from the user and sends multiple copies of it. This
+ * needs to be detected and the node charged for the duplicate transactions.
+ */
+/*@ThreadSafe*/
+public interface ReceiptCache {
+    /**
+     * Represents a single item in the cache.
+     * @param transactionID The transaction ID associated with the receipt
+     * @param nodeAccountIDs The list nodes which have independently submitted events with the same transaction ID
```

Comment:
I am a little confused. Here we are talking about nodes, that have **submitted** events with the same transaction ID. Below we record nodes that have **seen** a transaction ID. Is that the same thing?

Reply:
This will be redone in #6555